### PR TITLE
Implement lag feature for sql.go

### PR DESCRIPTION
**Pull request by Sami1309**
# Description

Implements functionality for lag features for creating training sets for sql offline providers. This will add columns based on feature columns in a training set shifted a certain increment back in time.

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: provider/sql.go [hunk 1]
```diff
@@ -1269,8 +1269,9 @@ func (q defaultOfflineSQLQueries) createValuePlaceholderString(columns []TableCo
 func (q defaultOfflineSQLQueries) trainingSetQuery(store *sqlOfflineStore, def TrainingSetDef, tableName string, labelName string, isUpdate bool) error {
 	columns := make([]string, 0)
 	query := ""
+	lagFeaturesOffset := 0
```
Comment (ahmadnazeri) [inline provider/sql.go:1272]: can we just move this down and do  `lagFeaturesOffset := len(def.Features)`

**Review by ahmadnazeri — APPROVED** [review]
LGTM

File: provider/sql.go (hunk 1 above)
Comment (Sami1309) [reply provider/sql.go]: smart thinking, changed

File: client/src/featureform/register.py [hunk 2]
```diff
@@ -1967,25 +1970,50 @@ def register_column_resources(
...
         for feature in features:
             if isinstance(feature, str):
                 feature_nv = (feature, "default")
                 feature_nv_list.append(feature_nv)
             elif isinstance(feature, dict):
-                feature_nv = (feature["name"], feature["variant"])
-                feature_nv_list.append(feature_nv)
+                lag = feature.get("lag")
+                if lag:
+                    if set(["lag", "feature", "variant"]) in set(feature.keys()):
```
Comment (sdreyer) [inline client/src/featureform/register.py:1981]: Should this be "not in"?

File: client/src/featureform/proto/metadata.proto [hunk 3]
```diff
@@ -5,6 +5,7 @@
 syntax = "proto3";
 
 import "google/protobuf/timestamp.proto";
+import "google/protobuf/duration.proto";
```
Comment (sdreyer) [inline client/src/featureform/proto/metadata.proto:8]: This file doesn't need to be edited. Its copied from metadata/proto

Comment (sdreyer) [reply client/src/featureform/proto/metadata.proto:8]: Should probably delete it from the repo

File: provider/spark.go [hunk 4]
```diff
@@ -175,6 +177,28 @@ func (q defaultSparkOfflineQueries) trainingSetCreate(def TrainingSetDef, featur
 		featureJoinQuery := fmt.Sprintf("LEFT OUTER JOIN (%s) t%d ON (t%d_entity = entity AND t%d_ts <= label_ts)", featureWindowQuery, i+1, i+1, i+1)
 		joinQueries = append(joinQueries, featureJoinQuery)
 	}
+	for i, lagFeature := range def.LagFeatures {
```
Comment (sdreyer) [inline provider/spark.go:180]: Is this what pandas is using?

Comment (ahmadnazeri) [reply provider/spark.go:180]: yeah, we are using the Spark SQL Queries for it. And the training set will be modified to include lag features.

**Pull request merged**